[Sbt over BSP] Extension points implementation

p-fuchs opened the pull request.

p-fuchs committed:
BSP plugin initial integration (#1)

Commit:
add a registry key for the bsp feature flag (#1)

Commit:
Merge pull request #2 from mapz-zpp/lamavia/feature-flag-registry-key

add a registry key for the bsp feature flag (#1)

Mikimm16 committed:
Added ScalaBspProjectAwareExtension. (#3)

* Added ScalaBspProjectAwareExtension.

* Removed automaticly added files.

---------

Co-authored-by: mikolaj <mikolajtuchola@gmail.com>

Commit:
LaMavia/bsp project open processor extension (#4)

* update id literal to a config constant (#1)

Signed-off-by: LaMavia <ts438730@students.mimuw.edu.pl>

* Update ScalaBspProjectOpenProcessor.scala

change build tool id to "sbt"

* Update ScalaBspProjectOpenProcessor.scala

remove legacy import

---------

Signed-off-by: LaMavia <ts438730@students.mimuw.edu.pl>

Commit:
add bsp integration icons (#6)

* add bsp integration icons

Signed-off-by: LaMavia <ts438730@students.mimuw.edu.pl>

* Update ScalaAssetsExtension.scala

change build tool id to "sbt"

---------

Signed-off-by: LaMavia <ts438730@students.mimuw.edu.pl>

Commit:
La mavia/bsp build target clasifier (#5)

* add SbtBuildTargetClassifier

Signed-off-by: LaMavia <ts438730@students.mimuw.edu.pl>

* fix extension error: change from object to class

Signed-off-by: LaMavia <ts438730@students.mimuw.edu.pl>

* rename to ScalaBuildTargetClassifier

Signed-off-by: LaMavia <ts438730@students.mimuw.edu.pl>

* update name to plugin config

Signed-off-by: LaMavia <ts438730@students.mimuw.edu.pl>

* Delete .idea/vcs.xml

* Update ImportsPanel.java

* Update ImportsPanel.java

* revert deletion

Signed-off-by: LaMavia <ts438730@students.mimuw.edu.pl>

* Update ImportsPanel.java

Change BorderFactory to IdeBorderFactory.PlainSmallWithIndent

---------

Signed-off-by: LaMavia <ts438730@students.mimuw.edu.pl>

p-fuchs committed:
Maintenance - new intellij-bsp and name changes (#8)

* Maintenance - new intellij-bsp and name changes

* Moved extension point

AnthonyGrod committed:
Anthony grod/bsp conn det gen ext (#7)

* Init basic methods in scalabspdetconngen

* add a registry key for the bsp feature flag (#1)

* Added ScalaBspProjectAwareExtension. (#3)

* Added ScalaBspProjectAwareExtension.

* Removed automaticly added files.

---------

Co-authored-by: mikolaj <mikolajtuchola@gmail.com>

* Implement all methods

* Change generateBspConnectionDetailsFile to scala-specific one

* Remove BspPluginTemplates file

* Remove coursier dependency

* Remove bsp folder

---------

Co-authored-by: agrodowski <agrodowski@iteratorshq.com>
Co-authored-by: Zuzanna Surowiec <z.surowiec@millionmonkeyssoftware.com>
Co-authored-by: Mikołaj Zakrzewski <95366924+Mikimm16@users.noreply.github.com>
Co-authored-by: mikolaj <mikolajtuchola@gmail.com>

p-fuchs commented on .idea/vcs.xml:
We constantly have problems with this file changing - it should not be changed, but could we postpone fixing it until final merge?

agluszak replied:
add .idea to .gitignore

agluszak commented on scala/integration/intellij-bsp/resources/META-INF/scala-intellij-bsp-integration.xml:
```diff
@@ -0,0 +1,12 @@
+        <bspConnectionDetailsGeneratorExtension
+                implementation="org.jetbrains.plugins.scala.bsp.flow.open.ScalaBspDetailsConnectionGenerator"/>
+    </extensions>
+</idea-plugin>
```
add newline

agluszak commented on scala/integration/intellij-bsp/src/org/jetbrains/plugins/scala/bsp/config/ScalaBspProjectAwareExtension.scala:
```diff
@@ -0,0 +1,18 @@
+import org.jetbrains.plugins.bsp.config.BspProjectAwareExtension
+import scala.jdk.CollectionConverters._
+
+import java.util
```
file is not formatted

agluszak replied:
I'd suggest using import name mapping instead

`import java.util.{List => JavaList}`

agluszak commented on scala/integration/intellij-bsp/src/org/jetbrains/plugins/scala/bsp/extension/ScalaBuildTargetClassifier.scala:
```diff
@@ -0,0 +1,34 @@
+private class ScalaBuildTargetClassifier extends BuildTargetClassifierExtension {
+  override val getBuildToolId: BuildToolId = ScalaPluginConstants.BUILD_TOOL_ID
+  override val getSeparator = "/"
+  private val sbtLabelRegex = """@?@?(?<repository>.*)//(?<package>.*):(?<target>.*)""".r
```
this looks like Bazel's labelling scheme, does it work for sbt too?

p-fuchs replied:
We should discuss at the tomorrow's meeting how this is presented in the UI to the user because right now we don't know what to look for.

agluszak commented on scala/integration/intellij-bsp/src/org/jetbrains/plugins/scala/bsp/flow/open/ScalaBspDetailsConnectionGenerator.scala:
```diff
@@ -0,0 +1,45 @@
+    getChild(projectPath, List(".bsp", "sbt.json").asJava)
+  }
+
+  override def calculateImportWizardSteps(path: Path, observableMutableProperty: ObservableMutableProperty[ConnectionFileOrNewConnection]): util.List[ImportProjectWizardStep] = ???
+
+  override def executeAndWait(list: util.List[String], virtualFile: VirtualFile, outputStream: OutputStream, project: Project): Unit = ???
```
these two are still not implemented

agluszak commented on scala/integration/intellij-bsp/src/org/jetbrains/plugins/scala/bsp/flow/open/ScalaBspDetailsConnectionGenerator.scala:
```diff
@@ -0,0 +1,45 @@
+
+  override def getChild(root: VirtualFile, path: util.List[String]): VirtualFile = {
+    val pathAsScala = path.asScala
+    val found = pathAsScala.foldLeft(root) { (vf, child) =>
+      vf.refresh(false, false)
+      vf.findChild(child)
+    }
+    found.refresh(false, false)
+    found
```
could you describe what it's supposed to do?

AnthonyGrod replied:
It's supposed to return a handle to a file pointed by the dependent `path`

agluszak commented on scala/integration/intellij-bsp/src/org/jetbrains/plugins/scala/project/bsp/ScalaSdkGetter.scala:
```diff
@@ -0,0 +1,46 @@
+      compilerBridgeBinaryJar = None
+    )
+  }
+}
```
add newline

p-fuchs committed:
Updated based on PR comments (#10)

* Addressed remarks from PR

* Reverted unused change

AnthonyGrod committed:
Fix generating bsp conn file (#11)

Co-authored-by: agrodowski <agrodowski@iteratorshq.com>

agluszak commented on scala/integration/intellij-bsp/src/org/jetbrains/plugins/scala/bsp/flow/open/ScalaBspDetailsConnectionGenerator.scala:
```diff
@@ -28,9 +29,26 @@ class ScalaBspDetailsConnectionGenerator extends BspConnectionDetailsGeneratorEx
   }
 
-  override def calculateImportWizardSteps(path: Path, observableMutableProperty: ObservableMutableProperty[ConnectionFileOrNewConnection]): util.List[ImportProjectWizardStep] = ???
+  override def calculateImportWizardSteps(path: Path, observableMutableProperty: ObservableMutableProperty[ConnectionFileOrNewConnection]): util.List[ImportProjectWizardStep] = List.empty.asJava
```
nit: use JavaList here as well (unless there's a different convention in intellij-scala)

Commit:
Upstream 2023.12.10 (#12)

* Add Scaladoc for the compilation request ordering implementation #SCL-21693

#SCL-CR-658

* Add reasoning behind the lack of compilation delay for incremental compilation requests #SCL-21693

#SCL-CR-658

* update sbt-idea-plugin to 3.22.0 (fallback to custom download url for plugin dependencies)

Main change we need from the new version
implement a mechanism to fallback to custom download url if the plugin is not found in marketplace

* Fix reentrancy problem in top level exports processing

* [bsp] improve logging for cancelled reload when BSP connection is forcefully stopped #SCL-21794

* [bsp] do not log process output if import process is finished #SCL-21794

Some implementation details.
To determine whether "Reload ALL BSP Projects" action should be active, it uses
`com.intellij.openapi.externalSystem.service.internal.ExternalSystemProcessingManager#hasTaskOfTypeInProgress`
(also see `com.intellij.openapi.externalSystem.action.RefreshAllExternalProjectsAction#update`)
`hasTaskOfTypeInProgress` checks if there are any running tasks by checking `myTasksInProgress` map.

When we force-kill the BSP connection
`com.intellij.openapi.externalSystem.service.internal.ExternalSystemProcessingManager#onEnd`
is invoked
from `com.intellij.openapi.externalSystem.service.internal.ExternalSystemResolveProjectTask#doExecute`
in "finally" section
![](image4.png){width=50%}

But after that, there is still some output coming from the input stream in `org.jetbrains.bsp.protocol.session.BspSession.BspProcessMessageHandler`
that output leads to invocation of `org.jetbrains.plugins.scala.build.ExternalSystemNotificationReporter#log`.
which leads to invocation of `com.intellij.openapi.externalSystem.service.internal.ExternalSystemProcessingManager#onTaskOutput`
`ExternalSystemProcessingManager#onTaskOutput` does this:
```java
myTasksInProgress.put(id, System.currentTimeMillis() + TOO_LONG_EXECUTION_MS);
```
Notice that it doesn't check  if `id` was present in the map.
But it was not, it was removed in `ExternalSystemProcessingManager#onEnd`
This leads to a stuck entry in the `myTasksInProgress` map

* [sbt] refactor: extract SbtShellCommunication.processNextQueuedCommand method #SCL-14720

* [sbt] notify pending commands in sbt shell communication queue about process termination #SCL-14720 fixed

* [sbt] add debug name to SbtShellReadyListener for easier debugging #SCL-14720

* reordered type inlay hint setting entries #SCL-21797

* fixed capitalization of range hints #SCL-21797

* fixed "show all type hints" action text #SCL-21797

* removed TypeMismatchToggleAction #SCL-21797

* fixed action texts #SCL-21797

* changed mismatch hint settings text #SCL-21797

* update intellijVersion to 233.11799.30

* context menu for type mismatch hints navigates to settings #SCL-21797 fixed

Because we set the filter and that is a text search, I updated the search index.

* Editor X-Ray mode: by-name argument hints, #SCL-16626

* Editor X-Ray mode: parameter hint pass, #SCL-16626

* Hint settings: misc

* Editor X-Ray mode: distinguish two Ctrl / Cmd keys, #SCL-16626

* Do not report silently added unused imports as errors in the Project View when there are `-Xfatal-warnings` enabled #SCL-21792

- Only when the project has both `-Xfatal-warnings` and `-Wunused-imports` on, will the unused imports be reported as errors.

* Update Project View error state independently of showing highlighting information in open editors #SCL-21792 fixed

- Additionally cancel all scheduled highlighting info application promises at the same time with only one `Cancellable` object, instead of allocating a cancellable object instance and an execution state object instance per promise.

* Add an optional timeout when waiting for the compile server to exit

- Override the timeout for ShowScalaCompilerTreeAction tests.

* Editor X-Ray mode: detailed type hint settings, #SCL-16626

* Editor X-Ray mode: activation settings simplified, #SCL-16626

* Editor X-Ray mode tip, #SCL-16626

* [find usages] place annotation usage in a correct usage type group #SCL-17348

* remove @ApiStatus.Internal annotations from AbstractTestRunConfiguration #SCL-21692 fixed

+ a couple of more classes, used in IntelliJ ZIO
See comment for reasoning:
https://youtrack.jetbrains.com/issue/SCL-21692/Please-remove-the-ApiStatus.Internal-annotation-from#focus=Comments-27-8384247.0-0

* Editor X-Ray mode: resource bundles, #SCL-16626

* Editor X-Ray mode: keyPressed repetition, #SCL-16626

* Factor out a CompilerHighlightingPsiChangeListener class in its own source file #SCL-21810

* Accept a virtual file instead of a file editor parameter when triggering incremental compilation #SCL-21810

* Use an editor focus changed listener instead of selected editor listener #SCL-21810

* Use a project activity for setting up compiler based highlighting in a project #SCL-21810

- This replaces the usage of a deprecated method in ProjectManagerListener.

* Register the focus change listener in an application level service #SCL-21810

- The FocusChangeListener interface listens to all editors from all projects. It should not be registered per project.
- We rely on the fact that services are instantiated once, so the focus change listener is also registered once.

* Fix the compiler highlighting tests to reflect the new trigger #SCL-21810 fixed

* [maven] when creating scala sdk fallback to test library dependency scope if we can't find it in compile scope #SCL-21701 fixed

* fixed toolwindow icons #SCL-21532 fixed

Mapping for toolwindow icons should reference a version without @20x20 because that should be automatically selected
Unfortunately getIcon does not load the correct @20x20 versions, but findIcon does. When #IDEA-338105 is fixed we might want to switch back to getIcon.
Also sbtShell@20x20.svg and sbtIcon@20x20.svg were completely different from their normal versions
in that they have no color in them. So I separated them into two different icons.
With all these changes, the toolwindow icon on the bottom left should now show
the bulky @20x20 version, while the icons in the View > ToolWindows menu should show the normal sleek version. Both without color.

* Add the Kotlin plugin as a runtime plugin #SCL-21501

* Editor X-Ray mode: widget, #SCL-16626

* Editor X-Ray mode: no widget in library sources, #SCL-16626

* Editor X-Ray mode: Cmd + click on widget, #SCL-16626

* Editor X-Ray mode: don't show in Find Action, #SCL-16626

* Editor X-Ray mode tip: show on Type Info, #SCL-16626

* Editor X-Ray mode: navigate to settings tab, #SCL-16626

* Editor X-Ray mode: navigation to parameters, #SCL-16626

* fixed indentation for then-branches when ALIGN_IF_ELSE is set #SCL-18488 fixed

* fixed parsing of type refinement for Scala 3 #SCL-21210 fixed

In scala 3 refinments are part of the simple type parsing rule.

* fixed return expr parsing #SCL-21823 fixed

* Editor X-Ray mode: custom tip for type hints, #SCL-16626

* Check if a virtual file is still valid before triggering CBH on file change #SCL-21820

* Do not blindly reschedule compilation requests #SCL-21820

- This change adds logic for determining if a request should be rescheduled or dropped.
- For example, requests for virtual files that are no longer valid are not rescheduled.
- Worksheet and document compilation requests need to have at least one editor open for a given document in order to be executed.

* Use the same compilation delay for all compilation requests

#SCL-21693 fixed
#SCL-21820

- Compilation requests which are not ready to be executed are delayed before scheduling again.
- The compilation delay will be surfaced as a project setting in a future commit.
- This change is necessary for further fixes, such as document compiler not being available before a successful incremental request.

* Do not enable the document compiler for a file before a successful incremental compilation request #SCL-21820

* Catch and log exceptions during execution of compilation requests #SCL-21820

* Editor X-Ray mode widget: resource bundle, #SCL-16626

* Update sbt-idea-plugin to version 3.23.1 and pick up fixes for adding the Kotlin plugin to the runtime classpath #SCL-21501 fixed

* [highlighting] pass PsiElement to HighlightInfo.Builder.range instead of TextRange, handle nulls #SCL-21796 fixed

HighlightInfo can be rejected by com.intellij.codeInsight.daemon.impl.HighlightInfoFilter

* Update intellijVersion to 233.11799.67

* [sbt] distinguish "resources" and "generated resources" roots #SCL-21826 fixed

* Change how Kotlin is detected in a project #SCL-21244

- Initially, we were trying to detect Kotlin libraries. At the moment, there doesn't seem to be a clear way how to detect a generic Kotlin SDK, at least not consistently across different project types (JPS, Maven, Gradle, sbt).
- With this change, we're now looking for Kotlin source files in the project. If they are detected, then IDEA is set as the incremental compiler.
- This change was made because projects with transitive dependencies on the Kotlin standard library (through dependencies which are written in Kotlin, or depend on code written in Kotlin), but which had no Kotlin sources were still being forced to use the IDEA incremental compiler.
- In the future, this workaround should no longer be necessary, when our Zinc integration is improved to use JPS fallback for modules which cannot be compiled by Zinc (that contain Kotlin or other languages).

* Editor X-Ray mode: scrolling, #SCL-16626

* fixed MacOS one pixel shift of aligned inlay hints #SCL-21630 fixed

the problem was that, macos has fractional pixel coordinates,
which pose a problem when using swings graphics api.
We were calculating the margin of the inlay hints in integers,
but that lead to rounding errors.
Now we just set the start of the inlay hints to the same x value for
all hints in an alignment-group instead of finding a margin
that correctly calculates this within the renderer

* Migrate the compiler highlighting focus listener from the platform's FocusChangeListener to Swing's more generic FocusAdapter in combination with an EditorFactoryListener #SCL-21810

- With a FocusChangeListener, there was an edge case where the focusLost event would not be fired for editors that are closed with the close button.
- This meant that the document compiler might not be disabled for a given file, and it would then be immediately recompiled with the document compiler the next time it was opened, without triggering incremental compilation first.
- With an EditorFactoryListener, we can intercept the event when the editor is closed, and disable the document compiler manually.

* Execute an incremental compilation request before compiling a worksheet for the first time #SCL-21838 fixed

* Show debugger value inlay hints for class parameters #SCL-19308 fixed

* Don't show parameter name hints by default, #SCL-16626

* Detect if the Kotlin plugin has been disabled when configuring the incremental compiler and skip changing the project settings #SCL-21244

* fix typo: anonimous -> anonymous in all safe places (comments, private members, add todos)

* [copy/paste] get common parent when there is a selection in editor

* [scala3 parser] unify ParserUtils.isSymbolicIdentifier implementation with Scala 3 implementation

* Editor X-Ray mode: only in Scala files, #SCL-16626

* Editor X-Ray mode: handle arrow keys, #SCL-16626

* Editor X-Ray mode: 500 ms key hold duration, #SCL-16626

* [scala3 resolve] resolve anonymous using parameters by name x$N (#SCL-21834, #SCL-21835) fixed

* fix text-to-text tests test data for scala 3 compiler sources

After the fixes in anonymous context parameter generated names

* Editor X-Ray mode: misc, #SCL-16626

* [copy/paste with bracelss syntax] fix pasting code in the beginning of a non-empty body #SCL-21664

* Use module scope when detecting Kotlin #SCL-21244

- This fixes a freeze caused by the seemingly huge content scope.

* Avoid duplication of module root changed events #SCL-21244

* Detect Kotlin source files in a background thread in a non blocking read action #SCL-21244

- Accessing the file indices and processing virtual files can be a slow operation. We do not want to block the UI.

* make scala 3 imported parser tests check psi tree #SCL-21539

* updated scala 3 imported parser testdata to have psi trees #SCL-21539

* fixed outdent for operators on the next line in scala 3 infix expr #SCL-21539 fixed

* assert some more stuff in imported parser tests #SCL-21832

* parse given signatures more gracefully #SCL-21832 fixed

* parse val/var in normal parameter clauses  #SCL-19151 fixed

* Add logging when a Kotlin source file is detected in a project #SCL-21244

* Editor X-Ray mode widget: hande Ctrl/Cmd clicks, #SCL-16626

* Editor X-Ray mode: stats, #SCL-16626

* fixed anonimous spelling error in imported parser tests #SCL-21539

* Hint folding: adjust mouse cursor, #SCL-21853 fixed

* [line commenter] respect "Add a space at line comment start" setting #SCL-21821 fixed

* Do not run the ScalaClassPostProcessorBuilder in non-Scala projects #SCL-21674

- In non-Scala projects, the builder was running and couldn't connect to the Scala compile server (expectedly), and it was printing warnings to the build window.

* fixed (parenthesized) int literal narrowing to byte/char/short alias #SCL-21842 fixed

* fixed fixed test :) #SCL-12611 fixed

* [formatter] better formatting for anonymous given instances/aliases #SCL-21832

* [formatter] add extra space between type and annotation #SCL-21860 fixed

Example: `val x: Int @unchecked = 42`

* Update intellijVersion to 233.11799.186

* Type hints: better navigation info, #SCL-21865 fixed

* pre-release-checklist: suppress Java highlighting in mixed Java and Scala 3 projects where Compiler Based Highlighting is enabled

#noport

* pre-release-checklist: disable "unused declaration" inspect in Scala 3 by default

#noport

* Type hints: navigation info context, #SCL-21865

* do not return null in ScGivenAliasDeclarationOrDefinitionImpl.nameId #SCL-21832

* moved test case into proper test #SCL-18488

* removed ASTNodeExt.isWhitespace #SCL-18488

* assert that icons are not null #SCL-21532

* pre-release-checklist: do not suppress Java highlighting in mixed Java and Scala 3 projects where Compiler Based Highlighting is enabled in nightly builds

#noport

* pre-release-checklist: enable "unused declaration" inspect in Scala 3 by default in nightly builds

#noport

* Hints: compute tooltip on demand, #SCL-21865

* [internal] updated IssueNavigationConfiguration in .idea/vcs.xml to recognise review patterns

+ some other helpful patterns

* [formatter] better formatting for anonymous given instances/aliases (part 2) #SCL-21832

SCL-CR-721

* Hint tooltips: fix possible NPE, #SCL-21865

* Uncomment test after parser improvements #SCL-19010 fixed

* Update intellijVersion to 233.11799.196

* Access the file indices during smart mode

#SCL-21869 fixed
#SCL-21244

- Detecting Kotlin source files requires accessing file indices, so the code should only be executed when the indices are available, i.e. in smart mode

* Do not try to apply highlighting information to the editor for files that have been removed while the compilation was still running #SCL-21870 fixed

* Add a default case when matching on boolean compile time ops #SCL-21875 fixed

- When matching on type parameters, a match error would be thrown. This is now fixed and covered by tests, for all other compile time ops as well.
- Only the boolean compile time ops had this problem.

Co-authored-by: Tobias Kahlert <tobias.kahlert@jetbrains.com>

* [completion] fix exhaustive match completion with indentation-based syntax #SCL-21714 fixed

* [structure view] implement sorting by visibility #SCL-13496 fixed

* [structure view] implement filtering out non-public elements #SCL-12875 fixed

* [structure view] extract sorters #SCL-CR-671, #SCL-CR-672

* [structure view] move filter to `filter` package

* [structure view] fix inherited variables and type aliases

* [structure view] show variables inherited from Java base class

* [structure view] show inherited classes

* [structure view] group inherited members by defining class #SCL-13552 fixed

* Only run the derives clause annotator when type aware annotations are enabled #SCL-21876 fixed

- We have minimized the actual underlying problem and I will open a follow-up ticket to fix the annotator itself.

Co-authored-by: Tobias Kahlert <tobias.kahlert@jetbrains.com>

* Editor X-Ray mode: method chain hints in Scala 3, #SCL-16626

* Rename kotlin to kotlinFileType for better code clarity

#SCL-21244

SCL-CR-700

* Factor out private methods instead of keeping the code in a large branch, hurting readability

#SCL-21244

SCL-CR-700

* migrate scala sdk extension impl from `intellij-bsp`

* Type hints: navigation info test, #SCL-21865

* Add a test for calculating source positions in the debugger

#SCL-19308

SCL-CR-703

* AmmoniteUnresolvedLibraryInspection: fix `NoSuchElementException: None.get` exception #SCL-21897 fixed

* [annotator] do not highlight unqualified renaming imports with Scala 3 syntax #SCL-21900 fixed

* Update intellijVersion to 233.11799.241

* Do not invoke Application#saveSettings as part of the Scala Compile Server start procedure on the UI thread

#SCL-21898 fixed

- First of all, the method which saves the settings itself (StoreUtil.saveSettings) is annotated with the `@CalledInAny` annotation. This allows the method to be called both from the UI thread and from background threads. The compile server start procedure runs on a background thread, so, we can call the saveSettings method from there, without having to move to the UI thread first.
- Secondly, moving the call to the UI thread and waiting for it to be executed (using invokeAndWait) can result in a deadlock in a very specific situation.
- When a project is opened, the platform runs an "is up-to-date check" using the JPS build system. We have a build listener which gets notified when this happens. The build listener ensures that the Scala Compile Server has been started, so it can be readily used for subsequent builds.
- If it happens that the user changes their mind and closes the project (or switches to another project by closing this one) while the Scala Compile Server is starting up, then IntelliJ IDEA will try to interrupt the up-to-date check. When this happens, IDEA will display a progress bar that cannot be interrupted. But this progress bar will also not allow the `saveSettings` method to be executed on the UI thread, because it will never be allowed to obtain the UI thread lock, resulting in a deadlock.

* Editor X-Ray mode: show even 1-char parameter hints, #SCL-21911

* [worksheet] filter out -Wunused:imports compiler option in worksheets #SCL-21880 fixed

* [worksheet] fix NPE in RunWorksheetActionPromoter.scala #EA-967526 fixed

* add test case for #SCL-21321

It seems #SCL-21321 was fixed by #SCL-21604.
In this change I just ensure that test case for unresolved object exists.

* [java decompiler] unify java decompiler with Java implementation #SCL-21903

- Use indentation size based on settings
- Decompile genetic signatures
- Show default constructor

* remove double bspIntegration from auto-resolve

* rename extension point from scalaSdkGetterExtension to scalaSdkExtension

---------

Co-authored-by: builduser <builduser@localhost>
Co-authored-by: Vasil Vasilev <vasil.vasilev@jetbrains.com>
Co-authored-by: Dmitrii Naumenko <dmitrii.naumenko@jetbrains.com>
Co-authored-by: Andrei Sugak <andrey.sugak@jetbrains.com>
Co-authored-by: Tobias Kahlert <tobias.kahlert@jetbrains.com>
Co-authored-by: Pavel Fatin <pavel.fatin@jetbrains.com>
Co-authored-by: Timur Aliberdov <timur.aliberdov@jetbrains.com>
Co-authored-by: Xuan-Son Trinh <xuanson.trinh@jetbrains.com>

jastice commented:
The current PR is huge, it might be due to including changes that happened on the base branch. I think you need to rebase your own changes on 233.x

agluszak commented:
> The current PR is huge, it might be due to including changes that happened on the base branch. I think you need to rebase your own changes on 233.x

to be clear: what Justin meant was updating your base branch (`no-idea`) and *then* rebasing this PR on that updated one
